libs/wire-subsystems/src/Wire/IndexedUserStore/ElasticSearch.hs

```diff
@@ -0,0 +1,161 @@
+import Data.ByteString.Conversion
+import Data.Id
+import Data.Text qualified as Text
+import Data.Text.Encoding qualified as Text
+import Database.Bloodhound qualified as ES
+import Imports
+import Network.HTTP.Client
+import Network.HTTP.Types
+import Polysemy
+import Wire.IndexedUserStore
+import Wire.Sem.Metrics (Metrics)
+import Wire.Sem.Metrics qualified as Metrics
+import Wire.UserSearch.Metrics
+import Wire.UserSearch.Types
+
+data ESConn = ESConn
+  { env :: ES.BHEnv,
+    indexName :: ES.IndexName
+  }
+
+data IndexedUserStoreConfig = IndexedUserStoreConfig
+  { conn :: ESConn,
+    additionalConn :: Maybe ESConn
+  }
+
+data IndexedUserStoreError
+  = IndexUpdateError ES.EsError
+  | IndexLookupError ES.EsError
+  | IndexError Text
+  deriving (Show)
+
+instance Exception IndexedUserStoreError
+
+interpretIndexedUserStoreES ::
+  ( Member (Embed IO) r,
+    Member Metrics r
+  ) =>
+  IndexedUserStoreConfig ->
+  InterpreterFor IndexedUserStore r
+interpretIndexedUserStoreES cfg =
+  interpret $ \case
+    Upsert docId userDoc versioning -> upsertImpl cfg docId userDoc versioning
+    UpdateTeamSearchVisibilityInbound tid vis -> updateTeamSearchVisibilityInboundImpl cfg tid vis
+    BulkUpsert docs -> bulkUpsertImpl cfg docs
+    DoesIndexExist -> doesIndexExistImpl cfg
+
+upsertImpl ::
+  forall r.
+  ( Member (Embed IO) r,
+    Member Metrics r
+  ) =>
+  IndexedUserStoreConfig ->
+  ES.DocId ->
+  UserDoc ->
+  ES.VersionControl ->
+  Sem r ()
+upsertImpl cfg docId userDoc versioning = do
+  void $ runInBothES cfg indexDoc
+  where
+    indexDoc :: ES.IndexName -> ES.BH (Sem r) ()
+    indexDoc idx = do
+      r <- ES.indexDocument idx mappingName settings userDoc docId
+      unless (ES.isSuccess r || ES.isVersionConflict r) $ do
+        lift $ Metrics.incCounter indexUpdateErrorCounter
+        res <- liftIO $ ES.parseEsResponse r
+        liftIO . throwIO . IndexUpdateError . either id id $ res
+      lift $ Metrics.incCounter indexUpdateSuccessCounter
+
+    settings = ES.defaultIndexDocumentSettings {ES.idsVersionControl = versioning}
+
+updateTeamSearchVisibilityInboundImpl :: forall r. (Member (Embed IO) r) => IndexedUserStoreConfig -> TeamId -> SearchVisibilityInbound -> Sem r ()
+updateTeamSearchVisibilityInboundImpl cfg tid vis =
+  void $ runInBothES cfg updateAllDocs
+  where
+    updateAllDocs :: ES.IndexName -> ES.BH (Sem r) ()
+    updateAllDocs idx = do
+      r <- ES.updateByQuery idx query (Just script)
+      unless (ES.isSuccess r || ES.isVersionConflict r) $ do
+        res <- liftIO $ ES.parseEsResponse r
+        liftIO . throwIO . IndexUpdateError . either id id $ res
+
+    query :: ES.Query
+    query = ES.TermQuery (ES.Term "team" $ idToText tid) Nothing
+
+    script :: ES.Script
+    script = ES.Script (Just (ES.ScriptLanguage "painless")) (Just (ES.ScriptInline scriptText)) Nothing Nothing
+
+    -- Unfortunately ES disallows updating ctx._version with a "Update By Query"
```

Comment:
Not updating versions can cause issues while reindexing. This seems like an oversight, we should try to break it and create tickets accordingly.

Reply:
This doesn't block this PR.